### PR TITLE
Custom kout/kneighbor, multi-node-shortest-path, jaccard-similar and template paths

### PR DESCRIPTION
**Pull request by zhoney**

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/algorithm/Consumers.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.function.Consumer;
+
+import org.slf4j.Logger;
+
+import com.baidu.hugegraph.HugeException;
+import com.baidu.hugegraph.task.TaskManager.ContextCallable;
+import com.baidu.hugegraph.util.ExecutorUtil;
+import com.baidu.hugegraph.util.Log;
+
+public class Consumers<V> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/algorithm/Consumers.java:37]: move Consumers to com/baidu/hugegraph/util

Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/job/algorithm/Consumers.java:37]: rebase https://github.com/hugegraph/hugegraph/pull/1183

**Review by houzhizhen — APPROVED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 2]
```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
...
+                    this.targetFinishOneStep = false;
+                }
             }
+        }
 
-            currentStep.decreaseTimes();
-            // Re-init sources
-            if (currentStep.remainTimes() > 0) {
-                this.sources = newVertices;
+        private void reInitCurrentIfNeeded(RepeatEdgeStep step,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:311]: reInitCurrentStepIfNeeded

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 3]
```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
...
-                this.sourcesAll.addAll(entry.getKey(), entry.getValue());
+                // Add node to next start-nodes
+                this.addNodeToNewVertices(target, new Node(target, n));
+            }
+        }
+
+        private void reInitAllIfNeeded(boolean forward) {
+            if (forward) {
+                // Re-init source all if last forward finished one super step
+                // and not last super step
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:296]: not finished last super step? or current step is not the last super step?
and use /*

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 4]
```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
...
+                if (this.sourceFinishOneStep && !this.lastSuperStep()) {
+                    this.sourcesAll = this.newMultiValueMap();
+                    this.sourceFinishOneStep = false;
+                }
+            } else {
+                // Re-init target all if last forward finished one super step
+                // and not last super step
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:303]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 5]
```diff
@@ -389,228 +413,89 @@ protected int accessedNodes() {
...
-            super(steps, capacity, limit, withRing);
-            for (Id id : sources) {
-                this.sources.add(id, new Node(id));
-            }
-            for (Id id : targets) {
-                this.targets.add(id, new Node(id));
-            }
-            this.paths = new PathSet();
+        public ConcurrentTraverser(Collection<Id> sources,
+                                   Collection<Id> targets,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:419]: move to previous line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 6]
```diff
@@ -389,228 +413,89 @@ protected int accessedNodes() {
...
-            // not last super step
-            if (this.targetFinishOneStep && !this.lastSuperStep()) {
-                this.targetsAll = newMultivalueMap();
-                this.targetFinishOneStep = false;
-            }
+        @Override
+        public void traverseOneLayer(
+                    Map<Id, List<Node>> vertices, RepeatEdgeStep step,
+                    BiConsumer<Id, RepeatEdgeStep> biConsumer) {
+            traverseIds(this.sources.keySet().iterator(), (id) -> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:434]: use this "this.traverseIds()"
and don't need "()" in "(id)"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 7]
```diff
@@ -389,228 +413,89 @@ protected int accessedNodes() {
...
-            // Re-init target all if last forward finished one super step and
-            // not last super step
-            if (this.targetFinishOneStep && !this.lastSuperStep()) {
-                this.targetsAll = newMultivalueMap();
-                this.targetFinishOneStep = false;
-            }
+        @Override
+        public void traverseOneLayer(
+                    Map<Id, List<Node>> vertices, RepeatEdgeStep step,
+                    BiConsumer<Id, RepeatEdgeStep> biConsumer) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:433]: just named consumer

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 8]
```diff
@@ -389,228 +413,89 @@ protected int accessedNodes() {
...
-            int size = 0;
-            for (List<Node> value : this.sourcesAll.values()) {
-                size += value.size();
-            }
-            for (List<Node> value : this.targetsAll.values()) {
-                size += value.size();
+        @Override
+        public void traverseOneLayer(
+                    Map<Id, List<Node>> vertices, RepeatEdgeStep step,
+                    BiConsumer<Id, RepeatEdgeStep> biConsumer) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:480]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 9]
```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
...
+            currentStep.swithDirection();
 
-        protected boolean finish() {
-            return this.stepCount >= this.totalSteps || this.reachLimit();
+            this.afterTraverse(currentStep, false);
         }
 
-        protected boolean lastStep() {
-            return this.stepCount == this.totalSteps - 1;
+        public RepeatEdgeStep step(boolean forward) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:187]: prefer nextStep

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 10]
```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
...
-            RepeatEdgeStep currentStep = null;
-            // Find next step to forward
-            for (int i = 0; i < this.steps.size(); i++) {
-                RepeatEdgeStep step = this.steps.get(i);
-                if (step.remainTimes() > 0) {
-                    currentStep = step;
-                    this.sourceIndex = i;
-                    break;
+        private void processOneForForward(Id source, Id target) {
+            for (Node n : this.sources.get(source)) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:240]: rename source to sourceV, and rename n to source, same as target

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 11]
```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
...
             }
+        }
 
-            // Traversal vertices of previous level
-            RepeatEdgeStep finalCurrentStep = currentStep;
-            traverseIds(this.sources.keySet().iterator(), vid -> {
-                if (this.reachLimit()) {
-                    return;
+        private void processOneForBackward(Id source, Id target) {
+            for (Node n : this.targets.get(source)) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:266]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java [hunk 12]
```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
...
             }
-            for (Id id : targets) {
-                this.targets.add(id, new Node(id));
+        }
+
+        private void processOne(Id source, Id target, boolean forward) {
+            if (forward) {
+                processOneForForward(source, target);
+            } else {
+                processOneForBackward(source, target);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:235]: can we merge the two methods with one, and pass sourcesAll or targetsAll to that method

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 13]
```diff
@@ -165,7 +165,7 @@ private HugeAuthenticator authenticator() {
 
     @SuppressWarnings("unused")
     private void installLicense(HugeConfig config, String md5) {
-        LicenseVerifier.instance().install(config, this, md5);
+//        LicenseVerifier.instance().install(config, this, md5);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:168]: keep origin

File: hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java [hunk 14]
```diff
@@ -286,4 +288,23 @@ public String writeWeightedPaths(WeightedPaths paths,
...
                                                "vertices", vertices));
     }
+
+    @Override
+    public String writeNodesWithPath(String name, Set<Id> nodes,
+                                     Collection<HugeTraverser.Path> paths,
+                                     Iterator<Vertex> iterator,
+                                     boolean countOnly) {
+        List<Map<String, Object>> pathList;
+        pathList = new ArrayList<>();
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/JsonSerializer.java:298]: keep in one line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java [hunk 15]
```diff
@@ -503,4 +503,21 @@ public static synchronized CoreOptions instance() {
...
+            );
+
+    public static final ConfigOption<Integer> OLTP_CONCURRENT_DEPTH =
+            new ConfigOption<>(
+                    "oltp.concurrent_depth",
+                    "The min depth to enable concurrent oltp algorithm.",
+                    rangeInt(0, 65535),
+                    10
+            );
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java:522]: remove empty line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java [hunk 16]
```diff
@@ -0,0 +1,587 @@
...
+        return nearest ? new SingleNearestTraverser(sources, targets, step,
+                                                    capacity, limit) :
+                         new SingleAllTraverser(sources, targets, step,
+                                                capacity, limit);
+    }
+
+    private Traverser concurrentTraverser(List<Id> sources, List<Id> targets,
+                                          EdgeStep step, boolean nearest,
+                                          long capacity, long limit) {
+        return new ConcurrentTraverser(sources, targets, step, capacity, limit);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java:111]: also refactor this class, and can we share code with templatepath?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CustomizedKneighborTraverser.java [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+            } else {
+                all.addAll(latest);
+            }
+        }
+
+        return all;
+    }
+
+    public Set<Node> customizedKneighborSingle(Id source, EdgeStep step,
+                                               int maxDepth, long limit) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CustomizedKneighborTraverser.java:86]: refactor with one method and call newSet() of Single and Concurrent

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CustomizedKoutTraverser.java [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+                // Update 'remaining' value to record remaining capacity
+                remaining -= latest.size();
+                reachCapacity(remaining, capacity, depth);
+            }
+        }
+
+        return latest;
+    }
+
+    public Set<Node> customizedKoutSingle(Id source, EdgeStep step,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CustomizedKoutTraverser.java:97]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/FusiformSimilarityTraverser.java [hunk 19]
```diff
@@ -194,7 +194,13 @@ public SimilarsMap fusiformSimilarity(Iterator<Vertex> vertices,
     }
 
     private static void checkGroupArgs(String groupProperty, int minGroups) {
-        if (groupProperty != null) {
+        if (groupProperty == null) {
+            E.checkArgument(minGroups == 0,
+                            "Can not set min group count when " +
+                                    "group property not set");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/FusiformSimilarityTraverser.java:200]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/FusiformSimilarityTraverser.java [hunk 20]
```diff
@@ -194,7 +194,13 @@ public SimilarsMap fusiformSimilarity(Iterator<Vertex> vertices,
     }
 
     private static void checkGroupArgs(String groupProperty, int minGroups) {
-        if (groupProperty != null) {
+        if (groupProperty == null) {
+            E.checkArgument(minGroups == 0,
+                            "Can not set min group count when " +
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/FusiformSimilarityTraverser.java:199]: Can't

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 21]
```diff
@@ -259,6 +262,35 @@ public double jaccardSimilarity(Id vertex, Id other, Directions dir,
         });
     }
 
+    protected Set<Node> adjacentVertices(Set<Node> vertices, EdgeStep step,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java:265]: move  Iterator<Id> adjacentVertices() to line 293

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 22]
```diff
@@ -217,10 +218,12 @@ public double jaccardSimilarity(Id vertex, Id other, Directions dir,
...
         Set<Id> targetNeighbors = IteratorUtils.set(this.adjacentVertices(
                                   other, dir, labelId, degree));
-        int interNum = CollectionUtil.intersect(sourceNeighbors,
-                                                targetNeighbors).size();
-        int unionNum = CollectionUtil.union(sourceNeighbors,
-                                            targetNeighbors).size();
+        return jaccardSimilarity(sourceNeighbors, targetNeighbors);
+    }
+
+    public double jaccardSimilarity(Set<Id> set1, Set<Id> set2) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java:224]: prefer move kout/jaccardSimilarity out of this class

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/JaccardSimilarTraverser.java [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+
+            double jaccardSimilarity = this.jaccardSimilarity(layer1s, layer3s);
+            results.put(id, jaccardSimilarity);
+            accessed.add(id);
+        });
+
+        return results;
+    }
+
+    public Map<Id, Double> jaccardSimilarsSingle(Id source, EdgeStep step,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/JaccardSimilarTraverser.java:119]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java (hunk 5 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java:419]: move to previous line will exceed 80 char

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java [hunk 24]
```diff
@@ -75,509 +73,215 @@ public CollectionPathsTraverser(HugeGraph graph) {
...
+
+        TraverseStrategy strategy;
+        if (depth >= this.concurrentDepth()) {
+            strategy = new ConcurrentTraverseStrategy(this.graph());
+            traverser = new Traverser(sourceList, targetList, step, depth,
+                                      capacity, limit, strategy);
+        } else {
+            strategy = new SingleTraverseStrategy(this.graph());
+            if (nearest) {
+                traverser = new SingleNearestTraverser(sourceList, targetList,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java:86]: no nearest when Concurrent?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java [hunk 25]
```diff
@@ -75,509 +73,215 @@ public CollectionPathsTraverser(HugeGraph graph) {
...
-        while (true) {
-            if (--depth < 0 || traverser.reachLimit()) {
-                break;
-            }
-            traverser.forward();
-
-            if (--depth < 0 || traverser.reachLimit()) {
-                break;
+        Traverser traverser;
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java:77]: unused empty line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 26]
```diff
@@ -540,7 +403,15 @@ public static void checkSkipDegree(long skipDegree, long degree,
     }
 
     protected static <V> Set<V> newSet() {
-        return new HashSet<>();
+        return newSet(true);
+    }
+
+    protected static <V> Set<V> newSet(boolean single) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java:409]: prefer rename to concurrent

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TpTraverser.java [hunk 27]
```diff
@@ -160,6 +177,199 @@ protected boolean match(Element elem, String key, Object value) {
         return p.isPresent() && Objects.equal(p.value(), value);
     }
 
+    protected TraverseStrategy traverseStrategy(boolean single) {
+        return single ? new SingleTraverseStrategy(this.graph()) :
+                        new ConcurrentTraverseStrategy(this.graph());
+
+    }
+
+    public abstract class PathTraverser {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TpTraverser.java:186]: add a new java file

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/strategy/ConcurrentTraverseStrategy.java [hunk 28]
```diff
@@ -71,4 +74,9 @@ public void addNewVerticesToAll(Map<Id, List<Node>> newVertices,
             vertices.addAll(entry.getKey(), entry.getValue());
         }
     }
+
+    @Override
+    public Iterator<Edge> edgesOfVertex(Id source, EdgeStep edgeStep) {
+        return super.edgesOfVertex(source, edgeStep);
+    }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/strategy/ConcurrentTraverseStrategy.java:81]: delete it

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/strategy/TraverseStrategy.java [hunk 29]
```diff
@@ -44,4 +47,6 @@ public abstract void addNode(Map<Id, List<HugeTraverser.Node>> vertices,
     public abstract void addNewVerticesToAll(
                          Map<Id, List<HugeTraverser.Node>> newVertices,
                          Map<Id, List<HugeTraverser.Node>> targets);
+
+    public Iterator<Edge> edgesOfVertex(Id source, EdgeStep edgeStep);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/strategy/TraverseStrategy.java:51]: delete it

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TpTraverser.java [hunk 30]
```diff
@@ -41,48 +38,34 @@
...
-import com.baidu.hugegraph.traversal.algorithm.strategy.TraverseStrategy;
 import com.baidu.hugegraph.util.Consumers;
 
 import jersey.repackaged.com.google.common.base.Objects;
 
 public abstract class TpTraverser extends HugeTraverser
                                   implements AutoCloseable {
 
-    protected static ExecutorService executor;
+    private static final String EXECUTOR_NAME = "oltp";
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TpTraverser.java:50]: also rename TpTraverser to OltpTraverser

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/strategy/TraverseStrategy.java [hunk 31]
```diff
@@ -48,5 +46,9 @@ public abstract void addNewVerticesToAll(
                          Map<Id, List<HugeTraverser.Node>> newVertices,
                          Map<Id, List<HugeTraverser.Node>> targets);
 
-    public Iterator<Edge> edgesOfVertex(Id source, EdgeStep edgeStep);
+    public static TraverseStrategy create(boolean concurrent, HugeGraph graph) {
+        return concurrent ? new ConcurrentTraverseStrategy(graph) :
+                            new SingleTraverseStrategy(graph);
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/strategy/TraverseStrategy.java:52]: empty line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TpTraverser.java [hunk 32]
```diff
@@ -41,48 +38,34 @@
...
         super(graph);
-    }
-
-    protected TpTraverser(HugeGraph graph, String name) {
-        super(graph);
         if (executor == null) {
             int workers = this.config().get(CoreOptions.OLTP_CONCURRENT_THREADS);
             if (workers > 0) {
-                executor = Consumers.newThreadPool(name, workers);
+                executor = Consumers.newThreadPool(EXECUTOR_NAME, workers);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TpTraverser.java:58]: lock for null judgment and assignment

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java (hunk 25 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CollectionPathsTraverser.java:72]: unused empty line

**Review by javeme — APPROVED** [review]

**Review by houzhizhen — APPROVED** [review]

**Review by Linary — APPROVED** [review]

**Pull request merged**